src/app/pages/data-protection/cloudsync/cloudsync-wizard/cloudsync-wizard.component.ts

```diff
@@ -0,0 +1,131 @@
+import { WebSocketService } from 'app/services/ws.service';
+
+@UntilDestroy()
+@Component({
+  templateUrl: './cloudsync-wizard.component.html',
+  styleUrls: ['./cloudsync-wizard.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  providers: [CloudCredentialService],
+})
+export class CloudsyncWizardComponent implements AfterViewInit {
+  @ViewChild(forwardRef(() => CloudsyncProviderComponent)) provider: CloudsyncProviderComponent;
+  @ViewChild(forwardRef(() => CloudsyncWhatAndWhenComponent)) whatAndWhen: CloudsyncWhatAndWhenComponent;
+  isLoading$ = new BehaviorSubject(false);
+  mergedLoading$: Observable<boolean>;
+
+  existingCredential: CloudsyncCredential;
+  createdProviders: CloudsyncCredential[] = [];
+  createdTasks: CloudSyncTask[] = [];
+
+  constructor(
+    public slideInRef: IxSlideInRef<CloudsyncWizardComponent>,
+    private ws: WebSocketService,
+    private snackbarService: SnackbarService,
+    private cdr: ChangeDetectorRef,
+    private translate: TranslateService,
+    private dialogService: DialogService,
+    private errorHandler: ErrorHandlerService,
+  ) {}
+
+  ngAfterViewInit(): void {
+    this.mergedLoading$ = merge(this.isLoading$, this.provider.isLoading$);
+    this.provider.form.controls.exist_credential.valueChanges
+      .pipe(filter(Boolean), untilDestroyed(this))
+      .subscribe((credential) => {
+        this.whatAndWhen?.form.patchValue({ credentials: credential });
+        this.cdr.markForCheck();
+      });
+
+    this.provider.form.controls.provider.valueChanges
+      .pipe(untilDestroyed(this))
+      .subscribe((provider) => {
+        const sourcePath = this.whatAndWhen?.form.controls.path_source.value.join(', ');
+        this.whatAndWhen?.form.patchValue({
+          description: `${cloudsyncProviderNameMap.get(provider)} - ${sourcePath}`,
+        });
+        this.cdr.markForCheck();
+      });
+  }
+
+  createTask(payload: CloudSyncTaskUpdate): Observable<CloudSyncTask> {
+    return this.ws.call('cloudsync.create', [payload]);
+  }
+
+  onProviderSaved(credential: CloudsyncCredential): void {
+    this.existingCredential = credential;
+    this.createdProviders.push(credential);
+    this.cdr.markForCheck();
+  }
+
+  onSubmit(): void {
+    this.isLoading$.next(true);
+
+    const payload = this.whatAndWhen.getPayload();
+
+    this.createTask(payload).pipe(
+      untilDestroyed(this),
+    ).subscribe({
+      next: (createdTask) => {
+        this.createdTasks.push(createdTask);
+        this.snackbarService.success(this.translate.instant('Task created'));
+        this.isLoading$.next(false);
+        this.slideInRef.close(true);
+        this.cdr.markForCheck();
+      }, error: (err) => {
+        this.dialogService.error(this.errorHandler.parseWsError(err));
+        this.rollBack();
+      },
+    });
+  }
+
+  rollBack(): void {
+    const requests: Observable<unknown>[] = [];
+
+    this.createdProviders.forEach((credential) => {
```

Comment:
Why are we making the assumption that if the task creation fails, the added credential must be deleted too. If the user wants to fix the bug and come back to add the task again, they would prefer to not go through the hassle of creating a provider again. Some providers are quite a hassle to create. You have to open another window, login into another service and provide permissions etc. So, we should be careful about rolling back credentials created.

Reply:
Good point. I agree. Let's keep providers for reusing.